# Table of Content
01. Import Libraries
02. Import Data
03. Change Column Data Type (if necessary)
04. Find Mixed-type Data
05. Data Accuracy
06. Data Consistency
07. Data Duplicates
08. Missing Values
09. Export Data

# 01. Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Adjust the setting to view all rows
pd.options.display.max_rows = None

In [3]:
# Adjust the setting to view all columns
pd.options.display.max_columns = None

# 02. Import Data

In [4]:
# Define the main project folder path
path = r'C:\Users\saich\Desktop\CareerFoundry\Data Immersion\Achievement 6 Advanced Analytics & Dashboard Design\11-2023 HDB Flat Resale Analysis'

In [5]:
# Import the 'MRT_LRT (combined)' sheet from 'train_station.xlsx' under 'Original Data' folder
# This file contains the opening and closure dates for all train stations (extracted from Wikipedia)
# https://en.wikipedia.org/wiki/List_of_Singapore_MRT_stations
# https://en.wikipedia.org/wiki/List_of_Singapore_LRT_stations

train_station = pd.read_excel(os.path.join(path, '02 Data', 'Original Data', 'train_station.xlsx'), 
                               sheet_name = 'MRT_LRT (combined)')

In [6]:
train_station

station_code        station_name                        line  \
0            NS1         Jurong East            North-South Line   
1            NS2         Bukit Batok            North-South Line   
2            NS3        Bukit Gombak            North-South Line   
3            NS4       Choa Chu Kang            North-South Line   
4            NS5             Yew Tee            North-South Line   
5            NS7              Kranji            North-South Line   
6            NS8           Marsiling            North-South Line   
7            NS9           Woodlands            North-South Line   
8           NS10           Admiralty            North-South Line   
9           NS11           Sembawang            North-South Line   
10          NS12            Canberra            North-South Line   
11          NS13              Yishun            North-South Line   
12          NS14              Khatib            North-South Line   
13          NS15        Yio Chu Kang            North-South Line   
14          NS16          Ang Mo Kio            North-South Line   
15          NS17              Bishan            North-South Line   
16          NS18            Braddell            North-South Line   
17          NS19           Toa Payoh            North-South Line   
18          NS20              Novena            North-South Line   
19          NS21              Newton            North-South Line   
20          NS22             Orchard            North-South Line   
21          NS23            Somerset            North-South Line   
22          NS24         Dhoby Ghaut            North-South Line   
23          NS25           City Hall            North-South Line   
24          NS26       Raffles Place            North-South Line   
25          NS27          Marina Bay            North-South Line   
26          NS28   Marina South Pier            North-South Line   
27           EW1           Pasir Ris              East-West Line   
28           EW2            Tampines              East-West Line   
29           EW3               Simei              East-West Line   
30           EW4         Tanah Merah              East-West Line   
31           EW5               Bedok              East-West Line   
32           EW6           Kembangan              East-West Line   
33           EW7               Eunos              East-West Line   
34           EW8          Paya Lebar              East-West Line   
35           EW9            Aljunied              East-West Line   
36          EW10             Kallang              East-West Line   
37          EW11            Lavender              East-West Line   
38          EW12               Bugis              East-West Line   
39          EW13           City Hall              East-West Line   
40          EW14       Raffles Place              East-West Line   
41          EW15       Tanjong Pagar              East-West Line   
42          EW16         Outram Park              East-West Line   
43          EW17         Tiong Bahru              East-West Line   
44          EW18             Redhill              East-West Line   
45          EW19          Queenstown              East-West Line   
46          EW20        Commonwealth              East-West Line   
47          EW21         Buona Vista              East-West Line   
48          EW22               Dover              East-West Line   
49          EW23            Clementi              East-West Line   
50          EW24         Jurong East              East-West Line   
51          EW25      Chinese Garden              East-West Line   
52          EW26            Lakeside              East-West Line   
53          EW27            Boon Lay              East-West Line   
54          EW28             Pioneer              East-West Line   
55          EW29            Joo Koon              East-West Line   
56          EW30          Gul Circle              East-West Line   
57          EW31       Tuas Crescent              East-West 

In [7]:
train_station.shape

(204, 5)

# 03. Change Column Data Type (if necessary)

In [8]:
# Check the data type of each column
train_station.dtypes

station_code    object
station_name    object
line            object
opening         object
closure         object
dtype: object

'Opening' and 'Closure' columns store date information. They should be changed into datetime data type. 

In [9]:
# Change the data types of 'opening' and 'closure' columns into datetime
train_station['opening'] = pd.to_datetime(train_station['opening'])

In [10]:
train_station['closure'] = pd.to_datetime(train_station['closure'])

In [11]:
train_station

station_code        station_name                        line    opening  \
0            NS1         Jurong East            North-South Line 1990-03-10   
1            NS2         Bukit Batok            North-South Line 1990-03-10   
2            NS3        Bukit Gombak            North-South Line 1990-03-10   
3            NS4       Choa Chu Kang            North-South Line 1990-03-10   
4            NS5             Yew Tee            North-South Line 1996-02-10   
5            NS7              Kranji            North-South Line 1996-02-10   
6            NS8           Marsiling            North-South Line 1996-02-10   
7            NS9           Woodlands            North-South Line 1996-02-10   
8           NS10           Admiralty            North-South Line 1996-02-10   
9           NS11           Sembawang            North-South Line 1996-02-10   
10          NS12            Canberra            North-South Line 2019-11-02   
11          NS13              Yishun            North-South Line 1988-12-20   
12          NS14              Khatib            North-South Line 1988-12-20   
13          NS15        Yio Chu Kang            North-South Line 1987-11-07   
14          NS16          Ang Mo Kio            North-South Line 1987-11-07   
15          NS17              Bishan            North-South Line 1987-11-07   
16          NS18            Braddell            North-South Line 1987-11-07   
17          NS19           Toa Payoh            North-South Line 1987-11-07   
18          NS20              Novena            North-South Line 1987-12-12   
19          NS21              Newton            North-South Line 1987-12-12   
20          NS22             Orchard            North-South Line 1987-12-12   
21          NS23            Somerset            North-South Line 1987-12-12   
22          NS24         Dhoby Ghaut            North-South Line 1987-12-12   
23          NS25           City Hall            North-South Line 1987-12-12   
24          NS26       Raffles Place            North-South Line 1987-12-12   
25          NS27          Marina Bay            North-South Line 1989-11-04   
26          NS28   Marina South Pier            North-South Line 2014-11-23   
27           EW1           Pasir Ris              East-West Line 1989-12-16   
28           EW2            Tampines              East-West Line 1989-12-16   
29           EW3               Simei              East-West Line 1989-12-16   
30           EW4         Tanah Merah              East-West Line 1989-11-04   
31           EW5               Bedok              East-West Line 1989-11-04   
32           EW6           Kembangan              East-West Line 1989-11-04   
33           EW7               Eunos              East-West Line 1989-11-04   
34           EW8          Paya Lebar              East-West Line 1989-11-04   
35           EW9            Aljunied              East-West Line 1989-11-04   
36          EW10             Kallang              East-West Line 1989-11-04   
37          EW11            Lavender              East-West Line 1989-11-04   
38          EW12               Bugis              East-West Line 1989-11-04   
39          EW13           City Hall              East-West Line 1987-12-12   
40          EW14       Raffles Place              East-West Line 1987-12-12   
41          EW15       Tanjong Pagar              East-West Line 1987-12-12   
42          EW16         Outram Park              East-West Line 1987-12-12   
43          EW17         Tiong Bahru              East-West Line 1988-03-12   
44          EW18             Redhill              East-West Line 1988-03-12   
45          EW19          Queenstown              East-West Line 1988-03-12   
46          EW20        Commonwealth              East-West Line 1988-03-12   
47          EW21         Buona Vista              East-West Line 1988-03-12   
48          EW22               Dover              East-West Line 2001-10-18   
49          EW23            Clementi              East

In [12]:
# Verify the changes
train_station.dtypes

station_code            object
station_name            object
line                    object
opening         datetime64[ns]
closure         datetime64[ns]
dtype: object

# 04. Find Mixed-type Data

In [13]:
# Check for any mixed-type columns - no output means no mixed-type data
for col in train_station.columns.tolist():
    weird = (train_station[[col]].applymap(type) != train_station[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (train_station[weird]) > 0:
        print (col)

closure


In [14]:
# Check the unique data types that the 'closure' column contains
# The apply() method in Pandas is used to apply a function along the axis of a DataFrame or Series. 
# In this case, the type function is applied to each element in the selected column. 

train_station['closure'].apply(type).unique()

array([<class 'pandas._libs.tslibs.nattype.NaTType'>,
       <class 'pandas._libs.tslibs.timestamps.Timestamp'>], dtype=object)

The 'closure' column contains a mix of datetime objects and 'NaT' values. This mix of data types triggers the detection of mixed-type data. Should not be an issue. 

In [15]:
train_station.applymap(type)

station_code   station_name           line  \
0    <class 'str'>  <class 'str'>  <class 'str'>   
1    <class 'str'>  <class 'str'>  <class 'str'>   
2    <class 'str'>  <class 'str'>  <class 'str'>   
3    <class 'str'>  <class 'str'>  <class 'str'>   
4    <class 'str'>  <class 'str'>  <class 'str'>   
5    <class 'str'>  <class 'str'>  <class 'str'>   
6    <class 'str'>  <class 'str'>  <class 'str'>   
7    <class 'str'>  <class 'str'>  <class 'str'>   
8    <class 'str'>  <class 'str'>  <class 'str'>   
9    <class 'str'>  <class 'str'>  <class 'str'>   
10   <class 'str'>  <class 'str'>  <class 'str'>   
11   <class 'str'>  <class 'str'>  <class 'str'>   
12   <class 'str'>  <class 'str'>  <class 'str'>   
13   <class 'str'>  <class 'str'>  <class 'str'>   
14   <class 'str'>  <class 'str'>  <class 'str'>   
15   <class 'str'>  <class 'str'>  <class 'str'>   
16   <class 'str'>  <class 'str'>  <class 'str'>   
17   <class 'str'>  <class 'str'>  <class 'str'>   
18   <class 'str'>  <class 'str'>  <class 'str'>   
19   <class 'str'>  <class 'str'>  <class 'str'>   
20   <class 'str'>  <class 'str'>  <class 'str'>   
21   <class 'str'>  <class 'str'>  <class 'str'>   
22   <class 'str'>  <class 'str'>  <class 'str'>   
23   <class 'str'>  <class 'str'>  <class 'str'>   
24   <class 'str'>  <class 'str'>  <class 'str'>   
25   <class 'str'>  <class 'str'>  <class 'str'>   
26   <class 'str'>  <class 'str'>  <class 'str'>   
27   <class 'str'>  <class 'str'>  <class 'str'>   
28   <class 'str'>  <class 'str'>  <class 'str'>   
29   <class 'str'>  <class 'str'>  <class 'str'>   
30   <class 'str'>  <class 'str'>  <class 'str'>   
31   <class 'str'>  <class 'str'>  <class 'str'>   
32   <class 'str'>  <class 'str'>  <class 'str'>   
33   <class 'str'>  <class 'str'>  <class 'str'>   
34   <class 'str'>  <class 'str'>  <class 'str'>   
35   <class 'str'>  <class 'str'>  <class 'str'>   
36   <class 'str'>  <class 'str'>  <class 'str'>   
37   <class 'str'>  <class 'str'>  <class 'str'>   
38   <class 'str'>  <class 'str'>  <class 'str'>   
39   <class 'str'>  <class 'str'>  <class 'str'>   
40   <class 'str'>  <class 'str'>  <class 'str'>   
41   <class 'str'>  <class 'str'>  <class 'str'>   
42   <class 'str'>  <class 'str'>  <class 'str'>   
43   <class 'str'>  <class 'str'>  <class 'str'>   
44   <class 'str'>  <class 'str'>  <class 'str'>   
45   <class 'str'>  <class 'str'>  <class 'str'>   
46   <class 'str'>  <class 'str'>  <class 'str'>   
47   <class 'str'>  <class 'str'>  <class 'str'>   
48   <class 'str'>  <class 'str'>  <class 'str'>   
49   <class 'str'>  <class 'str'>  <class 'str'>   
50   <class 'str'>  <class 'str'>  <class 'str'>   
51   <class 'str'>  <class 'str'>  <class 'str'>   
52   <class 'str'>  <class 'str'>  <class 'str'>   
53   <class 'str'>  <class 'str'>  <class 'str'>   
54   <class 'str'>  <class 'str'>  <class 'str'>   
55   <class 'str'>  <class 'str'>  <class 'str'>   
56   <class 'str'>  <class 'str'>  <class 'str'>   
57   <class 'str'>  <class 'str'>  <class 'str'>   
58   <class 'str'>  <class 'str'>  <class 'str'>   
59   <class 'str'>  <class 'str'>  <class 'str'>   
60   <class 'str'>  <class 'str'>  <class 'str'>   
61   <class 'str'>  <class 'str'>  <class 'str'>   
62   <class 'str'>  <class 'str'>  <class 'str'>   
63   <class 'str'>  <class 'str'>  <class 'str'>   
64   <class 'str'>  <class 'str'>  <class 'str'>   
65   <class 'str'>  <class 'str'>  <class 'str'>   
66   <class 'str'>  <class 'str'>  <class 'str'>   
67   <class 'str'>  <class 'str'>  <class 'str'>   
68   <class 'str'>  <class 'str'>  <class 'str'>   
69   <class 'str'>  <class 'str'>  <class 'str'>   
70   <class 'str'>  <class 'str'>  <class 'str'>   
71   <class 'str'>  <class 'str'>  <class 'str'>   
72   <class 'str'>  <class 'str'>  <class 'str'>   
73   <class 'str'>  <class 'str'>  <class 'str'>   
74   <class 'str'>  <class 'str'>  <class 'str'>   
75   <class 'str'>  <class 'str'>  <class 'str'>   
76

# 05. Data Accuracy

In [16]:
# Check the descriptive statistics for numerical data
train_station.describe()

opening              closure
count                            204                    1
mean   2005-12-23 01:38:49.411764736  2019-01-13 00:00:00
min              1987-11-07 00:00:00  2019-01-13 00:00:00
25%              1998-11-29 18:00:00  2019-01-13 00:00:00
50%              2008-07-07 12:00:00  2019-01-13 00:00:00
75%              2015-12-27 00:00:00  2019-01-13 00:00:00
max              2022-11-13 00:00:00  2019-01-13 00:00:00

No unusual findings.

# 06. Data Consistency

Check the data consistency for each categorical variable

### 06.1 station_code

In [17]:
# Check the unique values of 'Station code' column
train_station['station_code'].value_counts(dropna = False)

station_code
NS1     1
DT34    1
DT24    1
DT25    1
DT26    1
DT27    1
DT28    1
DT29    1
DT30    1
DT31    1
DT32    1
DT33    1
DT35    1
DT22    1
TE1     1
TE2     1
TE3     1
TE4     1
TE5     1
TE6     1
TE7     1
TE8     1
TE9     1
TE11    1
DT23    1
DT21    1
NS2     1
DT8     1
CC28    1
CC29    1
CE1     1
CE2     1
DT1     1
DT2     1
DT3     1
DT5     1
DT6     1
DT7     1
DT9     1
DT20    1
DT10    1
DT11    1
DT12    1
DT13    1
DT14    1
DT15    1
DT16    1
DT17    1
DT18    1
DT19    1
TE12    1
TE13    1
TE14    1
PE1     1
SE5     1
SW1     1
SW2     1
SW3     1
SW4     1
SW5     1
SW6     1
SW7     1
SW8     1
PTC     1
PE2     1
TE15    1
PE3     1
PE4     1
PE5     1
PE6     1
PE7     1
PW1     1
PW3     1
PW4     1
PW5     1
PW6     1
SE4     1
SE3     1
SE2     1
SE1     1
TE16    1
TE17    1
TE18    1
TE19    1
TE20    1
TE22    1
BP1     1
BP2     1
BP3     1
BP4     1
BP5     1
BP6     1
BP7     1
BP8     1
BP9     1
BP10    1
BP11    1
BP12    1
BP13   

No unusual findings. 

### 06.2 station_name

In [18]:
# Check the unique values of 'station_name' column
train_station['station_name'].value_counts(dropna = False)

station_name
Dhoby Ghaut           3
Marina Bay            3
Outram Park           3
Chinatown             2
Raffles Place         2
Tampines              2
Paya Lebar            2
Bugis                 2
Buona Vista           2
Expo                  2
HarbourFront          2
Jurong East           2
City Hall             2
Serangoon             2
Sengkang              2
Punggol               2
Promenade             2
MacPherson            2
Caldecott             2
Bayfront              2
Bukit Panjang         2
Stevens               2
Little India          2
Botanic Gardens       2
Orchard               2
Newton                2
Choa Chu Kang         2
Woodlands             2
Bishan                2
Somerset              1
Bright Hill           1
Woodlands North       1
Woodlands South       1
Springleaf            1
Lentor                1
Mayflower             1
Orchard Boulevard     1
Upper Thomson         1
Napier                1
Great World           1
Havelock              1
Max

No unusual findings. 

### 06.3 line

In [19]:
# Check the unique values of 'line' column
train_station['line'].value_counts(dropna = False)

line
Downtown Line                 34
East-West Line                33
Circle Line                   28
North-South Line              27
Thomsom-East Coast Line       20
North East Line               16
Bukit Panjang LRT             14
Sengkang LRT                  14
Punggol LRT                   14
Changi Airport Branch Line     2
Circle Line Extension          2
Name: count, dtype: int64

No unusual findings. 

# 07. Data Duplicates

In [20]:
# Check if there are any full duplicates in the dataframe
train_station.duplicated().sum()

0

In [21]:
train_station.loc[train_station.duplicated()]

Empty DataFrame
Columns: [station_code, station_name, line, opening, closure]
Index: []

No unusual findings.

# 08. Missing Values

In [22]:
# Check the number of missing values in each column
train_station.isnull().sum()

station_code      0
station_name      0
line              0
opening           0
closure         203
dtype: int64

There are 203 missing values in 'closure' column. 
This is because the train stations are still in operation, therefore they have no closure dates. 

# 09. Export Data

In [23]:
train_station.shape

(204, 5)

In [24]:
# Export data to 'Prepared Data' folder in csv format
train_station.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'train_station (checked).csv'))